File: java-to-kotlin/src/test/java/nl/rabobank/kotlinmovement/recipes/CreateUpdateRecipesControllerTest.java
package nl.rabobank.kotlinmovement.recipes;

import nl.rabobank.kotlinmovement.recipes.test.util.RecipeTest;
import nl.rabobank.kotlinmovement.recipes.test.util.model.RecipeRequestTest;
import nl.rabobank.kotlinmovement.recipes.test.util.model.RecipeResponseTest;
import nl.rabobank.kotlinmovement.recipes.test.util.model.RecipesErrorResponseTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.http.HttpMethod;

import java.util.stream.Stream;

import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeAssert.assertRecipeResponse;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.emptyRequest;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.emptyRequestIngredient;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.errorMessageIncorrectIngredientName;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.errorMessageIncorrectIngredientType;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.errorMessageIncorrectIngredients;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.errorMessageIncorrectRecipe;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.errorMessageIncorrectRecipeName;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.errorMessageIncorrectWeight;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.ingredientMissingName;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.ingredientMissingType;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.ingredientMissingWeight;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.nullIngredientsRequest;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.nullRecipeNameRequest;
import static nl.rabobank.kotlinmovement.recipes.test.util.RecipeTestData.peperoniPizzaRecipeRequest;
import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

class CreateUpdateRecipesControllerTest extends RecipeTest {

    @Test
    @DisplayName("Should be able to update a recipe")
    void test1() throws Exception {
        final RecipeRequestTest updateRequest = peperoniPizzaRecipeRequest;
        final RecipeResponseTest response = updateRecipe(1L, updateRequest);
        assertRecipeResponse(updateRequest, response);
    }

    @Test
    @DisplayName("Should be to able create new recipe when recipe id doesn't not exist")
    void test2() throws Exception {
        final RecipeRequestTest updateRequest = peperoniPizzaRecipeRequest;
        final RecipeResponseTest response = updateRecipe(2L, updateRequest);
        assertRecipeResponse(updateRequest, response);
    }

    @ParameterizedTest(name = "[{index}]: {2}")
    @MethodSource("errorDataParams")
    @DisplayName("Should not be able to create or update if request object is invalid")
    void test3(RecipeRequestTest recipeRequest, RecipesErrorResponseTest errorResponse, String testName) throws Exception {
        final String response = objectMapper.writeValueAsString(recipeRequest);
        final RecipesErrorResponseTest invalidCreateResponse = badRequestCall(HttpMethod.POST, "/recipes", response);
        final RecipesErrorResponseTest invalidUpdateResponse = badRequestCall(HttpMethod.PUT,"/recipes/1", response);
        assertThat(invalidCreateResponse).isEqualTo(errorResponse);
        assertThat(invalidUpdateResponse).isEqualTo(errorResponse);
    }

    private static Stream<Arguments> errorDataParams() {
        return Stream.of(
                Arguments.of(emptyRequest, errorMessageIncorrectRecipe, "create/update with empty RecipeRequest" ),
                Arguments.of(emptyRequestIngredient, errorMessageIncorrectIngredients, "create/update with empty RecipeRequest.ingredients"),
                Arguments.of(nullRecipeNameRequest, errorMessageIncorrectRecipeName, "create/update with missing RecipeRequest.recipeName field"),
                Arguments.of(nullIngredientsRequest, errorMessageIncorrectIngredients, "create/update with null for RecipeRequest.ingredients"),
                Arguments.of(ingredientMissingName, errorMessageIncorrectIngredientName, "create/update with missing IngredientRequest.name field"),
                Arguments.of(ingredientMissingType, errorMessageIncorrectIngredientType, "create/update with missing IngredientRequest.type field"),
                Arguments.of(ingredientMissingWeight, errorMessageIncorrectWeight, "create/update with missing IngredientRequest.weight field")
        );
    }

    @BeforeEach
    void setup() throws Exception {
        setInitialState(10);
    }

}

File: java-to-kotlin-complete/src/test/kotlin/nl/alimeshkat/recipes/CreateUpdateRecipesControllerTest.kt
package nl.alimeshkat.recipes

import nl.alimeshkat.recipes.test.util.RecipeAssert.assertRecipeResponse
import nl.alimeshkat.recipes.test.util.RecipeTest
import nl.alimeshkat.recipes.test.util.RecipeTestData.emptyRequest
import nl.alimeshkat.recipes.test.util.RecipeTestData.emptyRequestIngredient
import nl.alimeshkat.recipes.test.util.RecipeTestData.errorMessageIncorrectIngredientName
import nl.alimeshkat.recipes.test.util.RecipeTestData.errorMessageIncorrectIngredientType
import nl.alimeshkat.recipes.test.util.RecipeTestData.errorMessageIncorrectIngredients
import nl.alimeshkat.recipes.test.util.RecipeTestData.errorMessageIncorrectRecipe
import nl.alimeshkat.recipes.test.util.RecipeTestData.errorMessageIncorrectRecipeName
import nl.alimeshkat.recipes.test.util.RecipeTestData.errorMessageIncorrectWeight
import nl.alimeshkat.recipes.test.util.RecipeTestData.ingredientMissingName
import nl.alimeshkat.recipes.test.util.RecipeTestData.ingredientMissingType
import nl.alimeshkat.recipes.test.util.RecipeTestData.ingredientMissingWeight
import nl.alimeshkat.recipes.test.util.RecipeTestData.nullIngredientsRequest
import nl.alimeshkat.recipes.test.util.RecipeTestData.nullRecipeNameRequest
import nl.alimeshkat.recipes.test.util.RecipeTestData.peperoniPizzaRecipeRequest
import nl.alimeshkat.recipes.test.util.model.RecipeRequestTest
import nl.alimeshkat.recipes.test.util.model.RecipesErrorResponseTest
import org.assertj.core.api.AssertionsForClassTypes
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.http.HttpMethod
import java.util.stream.Stream

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
internal class CreateUpdateRecipesControllerTest : RecipeTest() {
    @Test
      fun `Should be able to update a recipe`()  {
        val updateRequest: RecipeRequestTest = peperoniPizzaRecipeRequest
        val response = updateRecipe(1L, updateRequest)
        assertRecipeResponse(updateRequest, response)
    }

    @Test
    fun `Should be to able create new recipe when recipe id doesn't not exist`() {
        val updateRequest: RecipeRequestTest = peperoniPizzaRecipeRequest
        val response = updateRecipe(2L, updateRequest)
        assertRecipeResponse(updateRequest, response)
    }

    @ParameterizedTest
    @MethodSource("errorDataParams")
    fun `Should not be able to create or update if request object is invalid`(
        recipeRequest: RecipeRequestTest?,
        errorResponse: RecipesErrorResponseTest?
    )  {
        val body = objectMapper.writeValueAsString(recipeRequest)
        val invalidCreateResponse = badRequestCall(HttpMethod.POST, "/recipes", body)
        val invalidUpdateResponse = badRequestCall(HttpMethod.PUT, "/recipes/1", body)
        AssertionsForClassTypes.assertThat(invalidCreateResponse).isEqualTo(errorResponse)
        AssertionsForClassTypes.assertThat(invalidUpdateResponse).isEqualTo(errorResponse)
    }

    @BeforeEach
      fun setup()  {
        setInitialState()
    }

    private fun errorDataParams(): Stream<Arguments> {
        return Stream.of(
            Arguments.of(emptyRequest, errorMessageIncorrectRecipe),
            Arguments.of(emptyRequestIngredient, errorMessageIncorrectIngredients),
            Arguments.of(nullRecipeNameRequest, errorMessageIncorrectRecipeName),
            Arguments.of(nullIngredientsRequest, errorMessageIncorrectIngredients),
            Arguments.of(ingredientMissingName, errorMessageIncorrectIngredientName),
            Arguments.of(ingredientMissingType, errorMessageIncorrectIngredientType),
            Arguments.of(ingredientMissingWeight, errorMessageIncorrectWeight)
        )
    }

}
